k8s/k8s.go

```diff
@@ -250,6 +263,118 @@ func CmdAddK8s(args *skel.CmdArgs, conf utils.NetConf, hostname string, calicoCl
+	// passed in, since CNI spec only supports one of each right now.
+	for _, ip := range ips {
+		ipAddr := net.ParseIP(ip)
+		if ipAddr == nil {
+			logger.WithField("IP", ip).Error("Invalid IP format")
+			return nil, fmt.Errorf("Invalid IP format: %s", ip)
+		}
+
+		// It's an IPv6 address if ip.To4 is nil.
+		if ipAddr.To4() == nil {
+			// We only allow one IPv4 and one IPv6 at the moment.
+			// So if we see more than one of IPv4 or IPv6 then we throw an error.
+			// If/when CNI spec supports more than one IP, we can loosen this requirement.
+			if result.IP6 != nil {
+				logger.Error("Can not have more than one IPv6 address in ipAddrs annotation")
+				return nil, fmt.Errorf("Can not have more than one IPv6 address in ipAddrs annotation")
+			}
+
+			// Calll callIPAMWithIP to set CNI_ARGS and call the IPAM plugin with it, and then unset it.
+			r, err := callIPAMWithIP(ipAddr, conf, args, logger)
+			if err != nil {
+				return nil, fmt.Errorf("Error getting IP from IPAM: %s", err)
+			}
+
+			// Set the IP6 part of the result from the type.Result returned by the IPAM plugin.
+			result.IP6 = r.IP6
+			logger.Debugf("Adding IPv6: %s to result", ipAddr.String())
+		} else {
+			// It's an IPv4 address.
+			if result.IP4 != nil {
+				logger.Error("Can not have more than one IPv4 address in ipAddrs annotation")
+				return nil, fmt.Errorf("Can not have more than one IPv4 address in ipAddrs annotation")
+			}
+
+			// Call callIPAMWithIP to set CNI_ARGS and call the IPAM plugin with it, and then unset it.
+			r, err := callIPAMWithIP(ipAddr, conf, args, logger)
+			if err != nil {
+				return nil, fmt.Errorf("Error getting IP from IPAM: %s", err)
+			}
+
+			// Set the IP4 part of the result from the type.Result returned by the IPAM plugin.
+			result.IP4 = r.IP4
+			logger.Debugf("Adding IPv4: %s to result", ipAddr.String())
+		}
+	}
+
+	return &result, nil
+}
+
+// callIPAMWithIP sets CNI_ARGS with the IP and calls the IPAM plugin with it
+// to get types.Result and then it unsets the IP field from CNI_ARGS ENV var,
+// so it doesn't pollute the subsequent requests.
+func callIPAMWithIP(ip net.IP, conf utils.NetConf, args *skel.CmdArgs, logger *log.Entry) (*types.Result, error) {
+
+	// Save the original value of the CNI_ARGS ENV var for backup.
+	originalArgs := os.Getenv("CNI_ARGS")
+	logger.Debugf("Original CNI_ARGS=%s", originalArgs)
+
+	// Append IP=ip_address to CNI_ARGS.
+	// See: https://github.com/containernetworking/cni/blob/master/CONVENTIONS.md#cni_args for more info.
+	newArgs := originalArgs + ";IP=" + ip.String()
+	logger.Debugf("New CNI_ARGS=%s", newArgs)
+
+	// Set CNI_ARGS to the new value.
+	err := os.Setenv("CNI_ARGS", newArgs)
+	if err != nil {
+		return nil, fmt.Errorf("Error setting CNI_ARGS environment variable: %v", err)
+	}
+
+	// Run the IPAM plugin.
+	logger.Debugf("Calling IPAM plugin %s", conf.IPAM.Type)
+	r, err := ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
```

Comment:
If we hit an error here, we don't unset the CNI_ARGS.

Might be best to use a `defer` so that we are guaranteed to reset the value.